crates/ntx-builder/src/actor/mod.rs

```diff
@@ -0,0 +1,339 @@
+
+/// The reason an actor has shut down.
+pub enum ActorShutdownReason {
+    /// Occurs when an account actor processes a transaction that reverts it's account creation.
+    AccountReverted(NetworkAccountPrefix),
+    /// Occurs when an account actor detects failure in the messaging channel used by the
+    /// coordinator.
+    EventChannelClosed,
+    /// Occurs when an account actor detects failure in acquiring the rate-limiting semaphore.
+    SemaphoreFailed(AcquireError),
+    /// Occurs when an account actor detects its corresponding cancellation token has been triggered
+    /// by the coordinator.
+    Cancelled(NetworkAccountPrefix),
+}
+
+// ACCOUNT ACTOR CONFIG
+// ================================================================================================
+
+/// Configuration that is required by all account actors.
+#[derive(Debug, Clone)]
+pub struct AccountActorConfig {
+    /// Client for interacting with the store in order to load account state.
+    pub store: StoreClient,
+    /// Address of the block producer gRPC server.
+    pub block_producer_url: Url,
+    /// Address of the remote prover. If `None`, transactions will be proven locally, which is
+    // undesirable due to the performance impact.
+    pub tx_prover_url: Option<Url>,
+    /// The latest chain state that account actors can rely on.
+    pub chain_state: Arc<RwLock<ChainState>>,
+}
+
+// ACCOUNT ORIGIN
+// ================================================================================================
+
+/// The origin of the account which the actor will use to initialize the account state.
+#[derive(Debug)]
+pub enum AccountOrigin {
+    /// Accounts that have been created by a transaction.
+    Transaction(Box<Account>),
+    /// Accounts that already exist in the store.
+    Store(NetworkAccountPrefix),
+}
+
+impl AccountOrigin {
+    /// Returns an [`AccountOrigin::Transaction`] if the account is a network account.
+    pub fn transaction(account: &Account) -> Option<Self> {
+        if account.is_network() {
+            Some(AccountOrigin::Transaction(account.clone().into()))
+        } else {
+            None
+        }
+    }
+
+    /// Returns an [`AccountOrigin::Store`].
+    pub fn store(prefix: NetworkAccountPrefix) -> Self {
+        AccountOrigin::Store(prefix)
+    }
+
+    /// Returns the [`NetworkAccountPrefix`] of the account.
+    pub fn prefix(&self) -> NetworkAccountPrefix {
+        match self {
+            AccountOrigin::Transaction(account) => NetworkAccountPrefix::try_from(account.id())
+                .expect("actor accounts are always network accounts"),
+            AccountOrigin::Store(prefix) => *prefix,
+        }
+    }
+}
+
+// ACTOR MODE
+// ================================================================================================
+
+/// The mode of operation that the account actor is currently performing.
+#[derive(Default, Debug)]
+enum ActorMode {
```

Comment:
Ideally, the states describe the current action as opposed to a property.

I'd propose `Idle`, `ProcessingNote` and `AwaitingTx`.